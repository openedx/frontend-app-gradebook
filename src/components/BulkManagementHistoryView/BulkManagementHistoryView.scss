.bulk-management-history-view {
  .help-text {
    margin-bottom: 40px;
    max-width: 70%;
  }
}
