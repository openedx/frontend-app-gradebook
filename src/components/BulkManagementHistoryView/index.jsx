/* eslint-disable react/button-has-type, import/no-named-as-default */
import React from 'react';
import { FormattedMessage } from '@edx/frontend-platform/i18n';

import messages from './messages';
import BulkManagementAlerts from './BulkManagementAlerts';
import HistoryTable from './HistoryTable';

/**
 * <BulkManagementHistoryView />
 * top-level view for managing uploads of bulk management override csvs.
 */
export const BulkManagementHistoryView = () => (
  <div className="bulk-management-history-view">
    <h4><FormattedMessage {...messages.heading} /></h4>
    <p className="help-text">
      <FormattedMessage {...messages.helpText} />
    </p>
    <BulkManagementAlerts />
    <HistoryTable />
  </div>
);

export default BulkManagementHistoryView;
