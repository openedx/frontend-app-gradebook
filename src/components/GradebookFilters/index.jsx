/* eslint-disable react/sort-comp, import/no-named-as-default */
import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import {
  Collapsible,
  Icon,
  IconButton,
  Form,
} from '@edx/paragon';
import { Close } from '@edx/paragon/icons';
import { FormattedMessage, injectIntl, intlShape } from '@edx/frontend-platform/i18n';

import actions from 'data/actions';
import selectors from 'data/selectors';
import thunkActions from 'data/thunkActions';

import messages from './messages';
import AssignmentTypeFilter from './AssignmentTypeFilter';
import AssignmentFilter from './AssignmentFilter';
import AssignmentGradeFilter from './AssignmentGradeFilter';
import CourseGradeFilter from './CourseGradeFilter';
import StudentGroupsFilter from './StudentGroupsFilter';

export class GradebookFilters extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      includeCourseRoleMembers: this.props.includeCourseRoleMembers,
    };
    this.handleIncludeTeamMembersChange = this.handleIncludeTeamMembersChange.bind(this);
  }

  handleIncludeTeamMembersChange(event) {
    const includeCourseRoleMembers = event.target.checked;
    this.setState({ includeCourseRoleMembers });
    this.props.updateIncludeCourseRoleMembers(includeCourseRoleMembers);
    this.props.fetchGrades();
    this.props.updateQueryParams({ includeCourseRoleMembers });
  }

  collapsibleGroup = (title, content) => (
    <Collapsible
      title={<FormattedMessage {...title} />}
      defaultOpen
      className="filter-group mb-3"
    >
      {content}
    </Collapsible>
  );

  render() {
    const {
      intl,
      updateQueryParams,
    } = this.props;
    return (
      <>
        <div className="filter-sidebar-header">
          <h2><Icon className="fa fa-filter" /></h2>
          <IconButton
            className="p-1"
            onClick={this.props.closeMenu}
            iconAs={Icon}
            src={Close}
            alt={intl.formatMessage(messages.closeFilters)}
            aria-label={intl.formatMessage(messages.closeFilters)}
          />
        </div>

        {this.collapsibleGroup(messages.assignments, (
          <div>
            <AssignmentTypeFilter updateQueryParams={updateQueryParams} />
            <AssignmentFilter updateQueryParams={updateQueryParams} />
            <AssignmentGradeFilter updateQueryParams={updateQueryParams} />
          </div>
        ))}

        {this.collapsibleGroup(messages.overallGrade, (
          <CourseGradeFilter updateQueryParams={updateQueryParams} />
        ))}

        {this.collapsibleGroup(messages.studentGroups, (
          <StudentGroupsFilter updateQueryParams={updateQueryParams} />
        ))}

        {this.collapsibleGroup(messages.includeCourseTeamMembers, (
          <Form.Checkbox
            checked={this.state.includeCourseRoleMembers}
            onChange={this.handleIncludeTeamMembersChange}
          >
            <FormattedMessage {...messages.includeCourseTeamMembers} />
          </Form.Checkbox>
        ))}
      </>
    );
  }
}
GradebookFilters.defaultProps = {
  includeCourseRoleMembers: false,
};
GradebookFilters.propTypes = {
  updateQueryParams: PropTypes.func.isRequired,
  // injected
  intl: intlShape.isRequired,
  // redux
  closeMenu: PropTypes.func.isRequired,
  fetchGrades: PropTypes.func.isRequired,
  includeCourseRoleMembers: PropTypes.bool,
  updateIncludeCourseRoleMembers: PropTypes.func.isRequired,
};

export const mapStateToProps = (state) => ({
  includeCourseRoleMembers: selectors.filters.includeCourseRoleMembers(state),
});

export const mapDispatchToProps = {
  closeMenu: thunkActions.app.filterMenu.close,
  fetchGrades: thunkActions.grades.fetchGrades,
  updateIncludeCourseRoleMembers: actions.filters.update.includeCourseRoleMembers,
};

export default injectIntl(connect(mapStateToProps, mapDispatchToProps)(GradebookFilters));
