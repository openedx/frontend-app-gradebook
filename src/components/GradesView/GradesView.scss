.spinner-overlay {
    position: fixed;
    height: 100%;
    width: 100%;
    top: 0;
    left: 0;
    background-color: #999;
    opacity: 0.5;
    z-index: 99999;
    display:flex;
    align-items: flex-start;
    justify-content: center;
    padding: 200px;
}

.color-black {
    color: black;
}

.gradebook-content {
    // note that this width isn't well-abstracted from Drawer code.
    // if we need to change it we may need to dig into those styles as well
    width: 100vw;
    .search-help-text {
        margin-left: 20px;
    }
    h4 {
        font-weight: bold;
        margin-top: 0rem;
    }
    .import-grades-btn {
      margin-left: 20px;
    }
    .intervention-report-description {
      margin-right: 40px;
    }
    h4.step-message-1 {
      margin-top: 40px;
    }
}

.student-filters{
    .label{
        padding-top: 30px;
    }
}
.grade-history-header{
    float: left;
}

.grade-history-assignment{
    padding-right: 49px;
}
.grade-history-student{
    padding-right: 75px;
}

.grade-history-original-grade{
    padding-right: 25px;
}

.grade-history-current-grade{
    padding-right: 25px;
}
.gradebook-container {
    width: 100%;
    overflow-x: auto;
    height: 600px;
    overflow-y: auto;
    position: relative;
}

.form-group, .pgn__form-group {
  label {
    font-weight: bold;
  }
}

.filter-group {
  .grade-filter-inputs {
    .percent-group {
      display: inline-block;
      .form-group, .pgn__form-group {
        width: 115px;
        display: inline-block;
      }
      .input-percent-label {
          margin-top: 22px;
          margin-left: 5px;
          margin-right: 5px;
      }
    }
  }
  .grade-filter-action {
    text-align: right;
  }
}

.mb-85 {
    margin-bottom: 85px;
}

.modal-dialog {
    max-width: 1000px;
}

.wrap-text-in-cell {
    overflow-wrap: break-word;
    word-wrap: break-word;
}

select#ScoreView.form-control {
  padding-right: 26px;
}

[dir=rtl] #course-grade-tooltip .arrow {
    right: initial;
    left: 0;
    
    &:before {
        border-width: 0.4rem 0.4rem 0.4rem 0;
        border-right-color: $black;
    }
}
