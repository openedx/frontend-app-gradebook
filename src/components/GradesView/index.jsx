/* eslint-disable react/sort-comp, react/button-has-type, import/no-named-as-default */
import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { FormattedMessage } from '@edx/frontend-platform/i18n';

import actions from 'data/actions';
import thunkActions from 'data/thunkActions';

import BulkManagementControls from './BulkManagementControls';
import EditModal from './EditModal';
import FilterBadges from './FilterBadges';
import FilteredUsersLabel from './FilteredUsersLabel';
import FilterMenuToggle from './FilterMenuToggle';
import GradebookTable from './GradebookTable';
import ImportSuccessToast from './ImportSuccessToast';
import InterventionsReport from './InterventionsReport';
import PageButtons from './PageButtons';
import ScoreViewInput from './ScoreViewInput';
import SearchControls from './SearchControls';
import SpinnerIcon from './SpinnerIcon';
import StatusAlerts from './StatusAlerts';
import messages from './messages';

export class GradesView extends React.Component {
  constructor(props) {
    super(props);
    this.handleFilterBadgeClose = this.handleFilterBadgeClose.bind(this);
  }

  handleFilterBadgeClose(filterNames) {
    return () => {
      this.props.resetFilters(filterNames);
      this.props.updateQueryParams(filterNames.reduce(
        (obj, filterName) => ({ ...obj, [filterName]: false }),
        {},
      ));
      this.props.fetchGrades();
    };
  }

  render() {
    return (
      <>
        <SpinnerIcon />

        <InterventionsReport />
        <h4 className="step-message-1">
          <FormattedMessage {...messages.filterStepHeading} />
        </h4>

        <div className="d-flex justify-content-between">
          <FilterMenuToggle />
          <SearchControls />
        </div>

        <FilterBadges handleClose={this.handleFilterBadgeClose} />
        <StatusAlerts />

        <h4><FormattedMessage {...messages.gradebookStepHeading} /></h4>

        <div className="d-flex justify-content-between align-items-center mb-2">
          <ScoreViewInput />
          <BulkManagementControls />
        </div>

        <FilteredUsersLabel />

        <GradebookTable />

        <PageButtons />
        <p>* <FormattedMessage {...messages.mastersHint} /></p>
        <EditModal />

        <ImportSuccessToast />
      </>
    );
  }
}

GradesView.defaultProps = {};

GradesView.propTypes = {
  updateQueryParams: PropTypes.func.isRequired,

  // redux
  fetchGrades: PropTypes.func.isRequired,
  resetFilters: PropTypes.func.isRequired,
};

export const mapStateToProps = () => ({});

export const mapDispatchToProps = {
  fetchGrades: thunkActions.grades.fetchGrades,
  resetFilters: actions.filters.reset,
};

export default connect(mapStateToProps, mapDispatchToProps)(GradesView);
