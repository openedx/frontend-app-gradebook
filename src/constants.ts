export const appId = 'org.openedx.frontend.app.gradebook';
