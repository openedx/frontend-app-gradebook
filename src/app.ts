import { App } from '@openedx/frontend-base';
import { appId } from './constants';
import routes from './routes';
import providers from './providers';
import messages from './i18n';
import slots from './slots';

const app: App = {
  appId,
  routes,
  providers,
  messages,
  slots,
  config: {
    LEARNING_BASE_URL: 'http://apps.local.openedx.io:2000',
    ENABLE_PROGRAMS: false,
    ECOMMERCE_BASE_URL: '',
    ORDER_HISTORY_URL: '',
    SUPPORT_URL: '',
  }
};

export default app;
