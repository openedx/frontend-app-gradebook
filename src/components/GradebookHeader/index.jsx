import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { getConfig } from '@edx/frontend-platform';
import { FormattedMessage } from '@edx/frontend-platform/i18n';
import { Button } from '@edx/paragon';

import { views } from 'data/constants/app';
import actions from 'data/actions';
import selectors from 'data/selectors';

import messages from './messages';

export class GradebookHeader extends React.Component {
  constructor(props) {
    super(props);
    this.handleToggleViewClick = this.handleToggleViewClick.bind(this);
  }

  get toggleViewMessage() {
    return this.props.activeView === views.grades
      ? messages.toActivityLog
      : messages.toGradesView;
  }

  lmsInstructorDashboardUrl = courseId => (
    `${getConfig().LMS_BASE_URL}/courses/${courseId}/instructor`
  );

  handleToggleViewClick() {
    const newView = this.props.activeView === views.grades ? views.bulkManagementHistory : views.grades;
    this.props.setView(newView);
  }

  render() {
    return (
      <div className="gradebook-header">
        <a
          href={this.lmsInstructorDashboardUrl(this.props.courseId)}
          className="mb-3"
        >
          <span aria-hidden="true">{'<< '}</span>
          <FormattedMessage {...messages.backToDashboard} />
        </a>
        <h1>
          <FormattedMessage {...messages.gradebook} />
        </h1>
        <div className="subtitle-row d-flex justify-content-between align-items-center">
          <h3>{this.props.courseId}</h3>
          { this.props.showBulkManagement && (
            <Button
              variant="tertiary"
              onClick={this.handleToggleViewClick}
            >
              <FormattedMessage {...this.toggleViewMessage} />
            </Button>
          )}
        </div>
        {this.props.areGradesFrozen
          && (
          <div className="alert alert-warning" role="alert">
            <FormattedMessage {...messages.frozenWarning} />
          </div>
          )}
        {(this.props.canUserViewGradebook === false) && (
          <div className="alert alert-warning" role="alert">
            <FormattedMessage {...messages.unauthorizedWarning} />
          </div>
        )}
      </div>
    );
  }
}

GradebookHeader.defaultProps = {
  // redux
  courseId: '',
  areGradesFrozen: false,
  canUserViewGradebook: false,
  showBulkManagement: false,
};

GradebookHeader.propTypes = {
  // redux
  activeView: PropTypes.string.isRequired,
  courseId: PropTypes.string,
  areGradesFrozen: PropTypes.bool,
  canUserViewGradebook: PropTypes.bool,
  setView: PropTypes.func.isRequired,
  showBulkManagement: PropTypes.bool,
};

export const mapStateToProps = (state) => ({
  activeView: selectors.app.activeView(state),
  courseId: selectors.app.courseId(state),
  areGradesFrozen: selectors.assignmentTypes.areGradesFrozen(state),
  canUserViewGradebook: selectors.roles.canUserViewGradebook(state),
  showBulkManagement: selectors.root.showBulkManagement(state),
});

export const mapDispatchToProps = {
  setView: actions.app.setView,
};

export default connect(mapStateToProps, mapDispatchToProps)(GradebookHeader);
