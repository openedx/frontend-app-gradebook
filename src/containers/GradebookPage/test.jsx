/* eslint-disable import/no-named-as-default */
import React from 'react';
import { shallow } from 'enzyme';
import queryString from 'query-string';

import selectors from 'data/selectors';
import thunkActions from 'data/thunkActions';

import GradebookFilters from 'components/GradebookFilters';
import GradebookHeader from 'components/GradebookHeader';
import GradesView from 'components/GradesView';
import BulkManagementHistoryView from 'components/BulkManagementHistoryView';
import { views } from 'data/constants/app';

import { GradebookPage, mapStateToProps, mapDispatchToProps } from '.';

jest.mock('query-string', () => ({
  parse: jest.fn(val => ({ parsed: val })),
  stringify: (val) => `stringify: ${JSON.stringify(val, Object.keys(val).sort())}`,
}));

jest.mock('@edx/paragon', () => ({
  Tab: () => 'Tab',
  Tabs: () => 'Tabs',
}));
jest.mock('data/selectors', () => ({
  __esModule: true,
  default: {
    app: {
      activeView: (state) => ({ activeView: state }),
    },
  },
}));
jest.mock('data/thunkActions', () => ({
  __esModule: true,
  default: {
    app: { initialize: jest.fn() },
  },
}));

jest.mock('components/WithSidebar', () => 'WithSidebar');
jest.mock('components/GradebookHeader', () => 'GradebookHeader');
jest.mock('components/GradesView', () => 'GradesView');
jest.mock('components/GradebookFilters', () => 'GradebookFilters');
jest.mock('components/BulkManagementHistoryView', () => 'BulkManagementHistoryView');

describe('GradebookPage', () => {
  describe('component', () => {
    const courseId = 'a course';
    let el;
    const props = {
      location: {
        search: 'searchString',
      },
      match: { params: { courseId } },
      activeView: views.grades,
    };
    beforeEach(() => {
      props.initializeApp = jest.fn();
      props.history = { push: jest.fn() };
    });
    test('snapshot - shows BulkManagementHistoryView if activeView === views.bulkManagementHistory', () => {
      el = shallow(<GradebookPage {...props} activeView={views.bulkManagementHistory} />);
      el.instance().updateQueryParams = jest.fn().mockName('updateQueryParams');
      expect(el.instance().render()).toMatchSnapshot();
    });
    test('snapshot - shows GradesView if aciveView === views.grades', () => {
      el = shallow(<GradebookPage {...props} />);
      el.instance().updateQueryParams = jest.fn().mockName('updateQueryParams');
      expect(el.instance().render()).toMatchSnapshot();
    });
    describe('render', () => {
      beforeEach(() => {
        el = shallow(<GradebookPage {...props} />);
      });
      describe('top-level WithSidebar', () => {
        test('sidebar from GradebookFilters, with updateQueryParams', () => {
          const { sidebar } = el.props();
          expect(sidebar).toEqual(
            <GradebookFilters updateQueryParams={el.instance().updateQueryParams} />,
          );
        });
      });
      describe('gradebook-content', () => {
        let content;
        let children;
        beforeEach(() => {
          content = el.props().children;
          children = content.props.children;
        });
        it('is wrapped in a div w/ px-3 gradebook-content classNames', () => {
          expect(content.type).toEqual('div');
          expect(content.props.className).toEqual('px-3 gradebook-content');
        });
        it('displays Gradebook header and then tabs', () => {
          expect(children[0]).toEqual(<GradebookHeader />);
        });
        it('displays GradesView if activeView === views.grades', () => {
          expect(shallow(children[1])).toEqual(shallow((
            <GradesView updateQueryParams={el.instance().updateQueryParams} />
          )));
        });
        it('displays Bulk Management History View if activeView === views.bulkManagementHistory', () => {
          el = shallow(<GradebookPage {...props} activeView={views.bulkManagementHistory} />);
          const mainView = el.props().children.props.children[1];
          expect(mainView).toEqual((
            <BulkManagementHistoryView />
          ));
        });
      });
    });
    describe('behavior', () => {
      beforeEach(() => {
        el = shallow(<GradebookPage {...props} />, { disableLifecucleMethods: true });
      });
      describe('componentDidMount', () => {
        test('initializes app with courseId and urlQuery', () => {
          el.instance().componentDidMount();
          expect(props.initializeApp).toHaveBeenCalledWith(
            courseId,
            queryString.parse(props.location.search),
          );
        });
      });
      describe('updateQueryParams', () => {
        it('replaces values for truthy values', () => {
          queryString.parse.mockImplementation(key => ({ [key]: key }));
          const newKey = 'testKey';
          const val1 = 'VALUE';
          const val2 = 'VALTWO!!';
          const args = { [newKey]: val1, [props.location.search]: val2 };
          el.instance().updateQueryParams(args);
          expect(props.history.push).toHaveBeenCalledWith(`?${queryString.stringify(args)}`);
        });
        it('clears values for non-truthy values', () => {
          queryString.parse.mockImplementation(key => ({ [key]: key }));
          const newKey = 'testKey';
          const val1 = 'VALUE';
          const val2 = false;
          const args = { [newKey]: val1, [props.location.search]: val2 };
          el.instance().updateQueryParams(args);
          expect(props.history.push).toHaveBeenCalledWith(
            `?${queryString.stringify({ [newKey]: val1 })}`,
          );
        });
      });
    });
  });
  describe('mapStateToProps', () => {
    let mapped;
    const testState = { trash: 'in', the: 'wind' };
    beforeEach(() => {
      mapped = mapStateToProps(testState);
    });
    test('activeView from app.activeView', () => {
      expect(mapped.activeView).toEqual(selectors.app.activeView(testState));
    });
  });
  describe('mapDispatchToProps', () => {
    test('initializeApp from thunkActions.app.initialize', () => {
      expect(mapDispatchToProps.initializeApp).toEqual(thunkActions.app.initialize);
    });
  });
});
