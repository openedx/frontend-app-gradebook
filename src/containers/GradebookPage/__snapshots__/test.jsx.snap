// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GradebookPage component snapshot - shows BulkManagementHistoryView if activeView === views.bulkManagementHistory 1`] = `
<WithSidebar
  sidebar={
    <GradebookFilters
      updateQueryParams={[MockFunction updateQueryParams]}
    />
  }
>
  <div
    className="px-3 gradebook-content"
  >
    <GradebookHeader />
    <BulkManagementHistoryView />
  </div>
</WithSidebar>
`;

exports[`GradebookPage component snapshot - shows GradesView if aciveView === views.grades 1`] = `
<WithSidebar
  sidebar={
    <GradebookFilters
      updateQueryParams={[MockFunction updateQueryParams]}
    />
  }
>
  <div
    className="px-3 gradebook-content"
  >
    <GradebookHeader />
    <GradesView
      updateQueryParams={[MockFunction updateQueryParams]}
    />
  </div>
</WithSidebar>
`;
